### DISCUSSION TITLE
Problems with chrootless setup and ZFS root

### DISCUSSION THREAD

**Discussion by danboid** [ltsp/ltsp#308] [Q&A]
I have installed the latest LTSP from the PPA under Ubuntu MATE 20.04 amd64 but I haven't been able to boot the LTSP image yet. When I try to boot from LTSP, I don't get much further than this PAM changes warning you can see in the attachments.

![IMG_20200914_225922](https://user-images.githubusercontent.com/1429783/93144450-eb102600-f6e1-11ea-928d-0c581e139458.jpg)

![IMG_20200914_225224](https://user-images.githubusercontent.com/1429783/93144460-ed728000-f6e1-11ea-9fd9-268e506a4197.jpg)

**Comment by alkisg** [comment]
Hi danboid,

https://github.com/ltsp/ltsp/issues is for bugs that can be solved by modifying the LTSP code.
Please use https://github.com/ltsp/community/issues instead for questions such as this one.

This message indicates that you updated the pam in your server without using `pam-auth-update`, which is what Debian and Ubuntu recommend for modifying pam. Please use `pam-auth-update` on the server, before running `ltsp image /`.

**Comment by danboid** [comment]
I presume I am having this problem because I chose to enable auto-login when I installed Ubuntu MATE?

I have run `pam-auth-update` but then I am asked which pam profiles I wish to enable or disable. I didn't think I wanted to change anything there so I just chose 'OK' and hoped that might fix it but it has not - I still get the pam warning when booting from LTSP.

I will try disabling auto-login and see if that fixes it.

**Comment by alkisg** [comment]
No, autologin is not an issue. You probably installed some package, like ldap or sssd, that messed up your pam configuration in an inappropriate manner.

What's the output of this command?

```shell
ls -lh /usr/share/pam-configs /etc/pam.d
```

**Comment by danboid** [comment]
```
$ ls -lh /usr/share/pam-configs/ /etc/pam.d/
/etc/pam.d/:
total 116K
-rw-r--r-- 1 root root  384 Feb  7  2020 chfn
-rw-r--r-- 1 root root   92 Feb  7  2020 chpasswd
-rw-r--r-- 1 root root  581 Feb  7  2020 chsh
-rw-r--r-- 1 root root 1.2K Sep 15 10:05 common-account
-rw-r--r-- 1 root root 1.3K Sep 15 10:05 common-auth
-rw-r--r-- 1 root root 1.5K Sep 15 10:05 common-password
-rw-r--r-- 1 root root 1.5K Sep 15 10:05 common-session
-rw-r--r-- 1 root root 1.5K Sep 15 10:05 common-session-noninteractive
-rw-r--r-- 1 root root  606 Feb 11  2020 cron
-rw-r--r-- 1 root root   69 Feb 17  2020 cups
-rw-r--r-- 1 root root  884 Aug 14  2019 lightdm
-rw-r--r-- 1 root root  551 Aug 14  2019 lightdm-autologin
-rw-r--r-- 1 root root  727 Aug 14  2019 lightdm-greeter
-rw-r--r-- 1 root root 4.1K Apr 16 13:36 login
-rw-r--r-- 1 root root   57 Sep 17  2018 mate-screensaver
-rw-r--r-- 1 root root   92 Feb  7  2020 newusers
-rw-r--r-- 1 root root  520 Dec 17  2019 other
-rw-r--r-- 1 root root   92 Feb  7  2020 passwd
-rw-r--r-- 1 root root  270 Aug 16  2019 polkit-1
-rw-r--r-- 1 root root  168 Feb  8  2019 ppp
-rw-r--r-- 1 root root  143 Jul 28  2019 runuser
-rw-r--r-- 1 root root  138 Jul 28  2019 runuser-l
-rw-r--r-- 1 root root   84 Jan  6  2020 samba
-rw-r--r-- 1 root root 2.1K May 29 08:37 sshd
-rw-r--r-- 1 root root 2.3K Jul 28  2019 su
-rw-r--r-- 1 root root  239 Feb  3  2020 sudo
-rw-r--r-- 1 root root  137 Jul 28  2019 su-l
-rw-r--r-- 1 root root  317 Apr 22 10:04 systemd-user

/usr/share/pam-configs/:
total 25K
-rw-r--r-- 1 root root 124 Feb 17  2020 capability
-rw-r--r-- 1 root root 152 Mar 11  2020 gnome-keyring
-rw-r--r-- 1 root root 154 Aug 24  2018 mkhomedir
-rw-r--r-- 1 root root 182 Jul  3 14:48 systemd
-rw-r--r-- 1 root root 682 Aug 24  2018 unix
```

**Comment by alkisg** [comment]
Can you also paste the output of `findmnt --real` on the server, and your ltsp.conf?

(Remove passwords etc if you have them in ltsp.conf)

**Comment by danboid** [comment]
```
$ findmnt --real
TARGET                       SOURCE                                           FSTYPE  OPTIONS
/                            rpool/ROOT/ubuntu_tqv0eh                         zfs     rw,relatime,xattr,posixacl
├─/sys/kernel/tracing        tracefs                                          tracefs rw,nosuid,nodev,noexec,relatime
├─/boot                      bpool/BOOT/ubuntu_tqv0eh                         zfs     rw,nodev,relatime,xattr,posixacl
│ ├─/boot/efi                /dev/sda1                                        vfat    rw,relatime,fmask=0022,dmask=0022,codepage=437,iocharset=iso8859-1,shortname=mixed,errors=remount-ro
│ └─/boot/grub               /dev/sda1[/grub]                                 vfat    rw,relatime,fmask=0022,dmask=0022,codepage=437,iocharset=iso8859-1,shortname=mixed,errors=remount-ro
├─/home/dan                  rpool/USERDATA/dan_vlkpo5                        zfs     rw,relatime,xattr,posixacl
├─/root                      rpool/USERDATA/root_vlkpo5                       zfs     rw,relatime,xattr,posixacl
├─/usr/local                 rpool/ROOT/ubuntu_tqv0eh/usr/local               zfs     rw,relatime,xattr,posixacl
├─/var/games                 rpool/ROOT/ubuntu_tqv0eh/var/games               zfs     rw,relatime,xattr,posixacl
├─/srv                       rpool/ROOT/ubuntu_tqv0eh/srv                     zfs     rw,relatime,xattr,posixacl
├─/var/mail                  rpool/ROOT/ubuntu_tqv0eh/var/mail                zfs     rw,relatime,xattr,posixacl
├─/var/log                   rpool/ROOT/ubuntu_tqv0eh/var/log                 zfs     rw,relatime,xattr,posixacl
├─/var/lib                   rpool/ROOT/ubuntu_tqv0eh/var/lib                 zfs     rw,relatime,xattr,posixacl
│ ├─/var/lib/AccountsService rpool/ROOT/ubuntu_tqv0eh/var/lib/AccountsService zfs     rw,relatime,xattr,posixacl
│ ├─/var/lib/apt             rpool/ROOT/ubuntu_tqv0eh/var/lib/apt             zfs     rw,relatime,xattr,posixacl
│ ├─/var/lib/NetworkManager  rpool/ROOT/ubuntu_tqv0eh/var/lib/NetworkManager  zfs     rw,relatime,xattr,posixacl
│ └─/var/lib/dpkg            rpool/ROOT/ubuntu_tqv0eh/var/lib/dpkg            zfs     rw,relatime,xattr,posixacl
├─/var/www                   rpool/ROOT/ubuntu_tqv0eh/var/www                 zfs     rw,relatime,xattr,posixacl
├─/var/spool                 rpool/ROOT/ubuntu_tqv0eh/var/spool               zfs     rw,relatime,xattr,posixacl
└─/var/snap                  rpool/ROOT/ubuntu_tqv0eh/var/snap                zfs     rw,relatime,xattr,posixacl
dan@dan-To-be-filled-by-O-E-M:~$ 
```

**Comment by danboid** [comment]
`/etc/ltsp/ltsp.conf` doesn't exist on my LTSP server.

**Comment by alkisg** [comment]
I guess that all these submounts do not exist at all in the client image, and pam fails because of the missing /var/lib.

You would need to put all these submounts in the ltsp image command like you did with /boot.

**Comment by danboid** [comment]
OK but that's a bit of a tedious thing to have to do.

I'm guessing from this that I'm the first person to try setting up LTSP on a Ubuntu machine with a ZFS root disk? It might also be worth noting this in the install guide, to say that ZFS root disks aren't fully supported by LTSP yet and to work around it you have to do this.

ZFS root installs are still an experimental feature in 20.04 so I can understand if you don't think its worth the effort in adding support for this yet.

**Comment by alkisg** [comment]
I haven't used ZFS, I do not know if all installations are affected or if it's something specific to your setup or even to the how the Ubuntu installer configures things.

Of course not all people use chrootless (`ltsp image /`); others may already be using ZFS servers with VMs or chroots, which wouldn't be affected by this issue.

Patches welcome; maybe a simple option like `mount --bind --some-other-option-to-also-mount-zfs-submounts` exists. Personally I won't be testing ZFS for at least a year.

**Comment by alkisg** [comment]
LTSP essentially tries something like the following:

```shell
mkdir /run/tmp
mount --bind / /run/tmp
```

Then, check inside /run/tmp on what's missing; if all the submounts are missing, that's the issue. If you find a parameter that properly mounts them (but excludes /dev and all the other mounts that shouldn't be replicated), that would be the basis of a patch.

**Comment by danboid** [comment]
I'd recommend you check out ZFS as soon as possible. It's quite a lot to learn but its well worth it. I use ZFS wherever I can and have done so for years now.

The key things to check out are boot environments / grub history under 20.4 when zsys is installed and snapshots. zsys creates new grub boot environments every time you install or remove a package or do an update, or you can create them manually ofc.

You'll wonder how you lived without them after you've seen how powerful they are. Instant recoveries! You can have 10s of thousands of snapshots on a dataset a ZFS chugs along happily.

**Comment by danboid** [comment]
Proxmox VE currently has the best ZFS support that I've seen in a Linux distro as it easily lets you create and install to RAIDZ2/Z3 pools. The Ubuntu installer current gives you exactly 0 options for ZFS.
